**Question 1** : Créer une SparkSession avec le nom d'application 'TradeCorp ETL'. Afficher la version de Spark.

In [4]:
import logging
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("TradeCorpETL") \
    .getOrCreate()

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s | %(levelname)s | %(message)s"
)

logger = logging.getLogger("TradeCorpETL")

spark = SparkSession.builder.appName("TradeCorpETL").getOrCreate()

logger.info(f"Session Spark créée - version {spark.version}")

2026-08-31 10:11:26,852 | INFO | Session Spark créée - version 4.2.0


**Question 2** : Lire les 8 fichiers CSV. Stocker chaque DataFrame dans une variable.


In [13]:
logger.info("Question 2 - stocker chaque DataFrame dans une variable")
files = [
    "customers",
    "orders",
    "order_details",
    "products",
    "categories",
    "suppliers",
    "employees",
    "shippers"
]

dataframes = {}

for file in files:
    logger.info(f"Chargement de {file}.csv")

    dataframes[file] = spark.read.csv(
        f"../data/{file}.csv",
        header=True,
        inferSchema=True
    )

    logger.info(
        f"{file}.csv chargé avec succès : "
        f"{dataframes[file].count()} lignes"
    )

logger.info("Question 2 terminée")

2026-08-31 10:18:23,771 | INFO | Question 2 - stocker chaque DataFrame dans une variable
2026-08-31 10:18:23,773 | INFO | Chargement de customers.csv
2026-08-31 10:18:24,029 | INFO | customers.csv chargé avec succès : 91 lignes
2026-08-31 10:18:24,029 | INFO | Chargement de orders.csv
2026-08-31 10:18:24,279 | INFO | orders.csv chargé avec succès : 830 lignes
2026-08-31 10:18:24,280 | INFO | Chargement de order_details.csv
2026-08-31 10:18:24,563 | INFO | order_details.csv chargé avec succès : 2155 lignes
2026-08-31 10:18:24,565 | INFO | Chargement de products.csv
2026-08-31 10:18:24,800 | INFO | products.csv chargé avec succès : 77 lignes
2026-08-31 10:18:24,801 | INFO | Chargement de categories.csv
2026-08-31 10:18:25,023 | INFO | categories.csv chargé avec succès : 8 lignes
2026-08-31 10:18:25,024 | INFO | Chargement de suppliers.csv
2026-08-31 10:18:25,292 | INFO | suppliers.csv chargé avec succès : 29 lignes
2026-08-31 10:18:25,293 | INFO | Chargement de employees.csv
2026-08-31 1

**Question 3** : Pour chaque DataFrame, afficher le schema des données. Identifier les types de colonnes inférés
automatiquement.

In [11]:
logger.info("Question 3 - Affichage des schémas des DataFrames")

for name, df in dataframes.items():
    print(f"{name}")
    df.printSchema()

logger.info("Affichage des schémas terminé")
logger.info("Question 3 terminée")

2026-08-31 10:14:22,215 | INFO | Question 3 - Affichage des schémas des DataFrames
2026-08-31 10:14:22,222 | INFO | Affichage des schémas terminé
2026-08-31 10:14:22,223 | INFO | Question 3 terminée


customers
root
 |-- customer_id: string (nullable = true)
 |-- company_name: string (nullable = true)
 |-- contact_name: string (nullable = true)
 |-- contact_title: string (nullable = true)
 |-- address: string (nullable = true)
 |-- city: string (nullable = true)
 |-- region: string (nullable = true)
 |-- postal_code: string (nullable = true)
 |-- country: string (nullable = true)
 |-- phone: string (nullable = true)
 |-- fax: string (nullable = true)

orders
root
 |-- order_id: integer (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- employee_id: integer (nullable = true)
 |-- order_date: date (nullable = true)
 |-- required_date: date (nullable = true)
 |-- shipped_date: date (nullable = true)
 |-- ship_via: integer (nullable = true)
 |-- freight: double (nullable = true)
 |-- ship_name: string (nullable = true)
 |-- ship_address: string (nullable = true)
 |-- ship_city: string (nullable = true)
 |-- ship_region: string (nullable = true)
 |-- ship_postal_code: stri

Globalement les colonnes sont bien inférés par Spark, cependant certaines sont mise en string comme quantity_per_unit dans products alors qu'un int est plus judicieux. 

**Question 4** : Pour chaque DataFrame, afficher les 5 premières lignes. Observer la structure des données.


In [16]:
logger.info("Question 4 - Affichage des 5 premières lignes des DataFrames")

for name, df in dataframes.items():
    print(f"\n{name} : ")
    df.show(5)

2026-08-31 10:20:17,937 | INFO | Question 4 - Affichage des 5 premières lignes des DataFrames



customers : 
+-----------+--------------------+------------------+--------------------+--------------------+-----------+------+-----------+-------+--------------+--------------+
|customer_id|        company_name|      contact_name|       contact_title|             address|       city|region|postal_code|country|         phone|           fax|
+-----------+--------------------+------------------+--------------------+--------------------+-----------+------+-----------+-------+--------------+--------------+
|      ALFKI| Alfreds Futterkiste|      Maria Anders|Sales Representative|       Obere Str. 57|     Berlin|  NULL|      12209|Germany|   030-0074321|   030-0076545|
|      ANATR|Ana Trujillo Empa...|      Ana Trujillo|               Owner|Avda. de la Const...|México D.F.|  NULL|      05021| Mexico|  (5) 555-4729|  (5) 555-3745|
|      ANTON|Antonio Moreno Ta...|    Antonio Moreno|               Owner|     Mataderos  2312|México D.F.|  NULL|      05023| Mexico|  (5) 555-3932|          NU

**Question 5** : Compter le nombre de lignes de chaque DataFrame et créer un tableau récapitulatif

In [15]:
logger.info("Question 5 - Compter le nombre de lignes dans chaque DF")

for name, df in dataframes.items():
    print(f"{name} : {df.count()} lignes")

2026-08-31 10:20:10,766 | INFO | Question 5 - Compter le nombre de lignes dans chaque DF


customers : 91 lignes
orders : 830 lignes
order_details : 2155 lignes
products : 77 lignes
categories : 8 lignes
suppliers : 29 lignes
employees : 9 lignes
shippers : 6 lignes


**Question 6** : Sur df_orders et df_products essayer d’obtenir les statistiques suivantes : min, max, mean, stddev

In [29]:
df_orders.describe().show()
df_products.describe().show()

+-------+-----------------+-----------+------------------+------------------+------------------+--------------------+--------------------+---------+-----------+------------------+------------+
|summary|         order_id|customer_id|       employee_id|          ship_via|           freight|           ship_name|        ship_address|ship_city|ship_region|  ship_postal_code|ship_country|
+-------+-----------------+-----------+------------------+------------------+------------------+--------------------+--------------------+---------+-----------+------------------+------------+
|  count|              830|        830|               830|               830|               830|                 830|                 830|      830|        323|               811|         830|
|   mean|          10662.5|       NULL| 4.403614457831325|2.0072289156626506| 78.24420481927719|                NULL|                NULL|     NULL|       NULL|39975.067357512955|        NULL|
| stddev|239.7446558319914|       N

**Question 7** : Expliquer dans une cellule Markdown ce qu'est la lazy evaluation dans Spark. Quelle est la différence entre une
transformation et une action ? Donner 3 exemples de chaque

La lazy evaluation (évaluation paresseuse) signifie que lorsqu'on applique des trasnformations sur un DataFrame, Spark ne les exécute pas immédiatement.
Les traitements sont exécutés uniquement lorsqu'un action demande un résultat.

### Exemple de transformation
Une transformation va créer un nouveau DataFrame à partir de l'existant, sans déclancher immédiatement le calcul. 
- `filter()` : Filtre les lignes
- `distinct()` : retire les lignes identiques
- `groupBy()` : prépare un regroupement pour une agrégation

### Exemple d'action
Une action déclenche l'exécution des transformations définies afin d'obtenir un résultat
- `show()` : Afficher les données
- `count()` : Retourner le nombre de lignes
- `head(5)` : récupèrer les cinq premières lignes

Par exemple :

`df.filter(df.country == "France")`

ça ne déclenche pas immédiatement le traitement. Spark attend qu'une action soit appelée.

`df.filter(df.country == "France").show()`

C'est l'appel à l'action `show()` qui déclenche réellement l'exécution du traitement.

**Question 8** : Ouvrir http://localhost:4040 dans le navigateur. Observer les jobs exécutés. Identifier ce que représentent les
stages et tasks.

Les jobs représentent les actions spark (par exemple je peux voir les count(), ou encore les show() que j'ai pu faire.)

Un job est découpé en plusieurs étapes d'exécution, c'est les stages. Un nouveau stage est générélement crée lorsque Spark doit redistribuer
les données entre les partitions (shuffle)

Une task est une plus petite unité de travail d'un stage. Chaque task traite une partition des données
et plusieurs tasks peuvent être exécutées en parallèle. 

**Question 9** : Lire le fichier orders.csv avec Pandas et mesurer le temps. Puis avec Spark. Comparer. Expliquer quand Spark
est plus pertinent que Pandas.

In [18]:
import pandas as pd
import time

start = time.perf_counter()

df_pandas = pd.read_csv("../data/orders.csv")

end = time.perf_counter()

logger.info(f"Temps Pandas : {end - start:.4f} secondes")

start = time.perf_counter()

df_spark = spark.read.csv("../data/orders.csv", header=True, inferSchema=True)

df_spark.count()

end = time.perf_counter()

logger.info(f"Temps Spark : {end - start:.4f} secondes")

2026-08-31 10:21:03,859 | INFO | Temps Pandas : 0.0098 secondes
2026-08-31 10:21:04,094 | INFO | Temps Spark : 0.2343 secondes


Sur ce fichier, Pandas est plus rapide que Spark car le dataset est petit, il ne fait que 830 lignes.

Pandas est plus adapté aux petits volumes de données pouvant être traités en mémoire sur une seule machine.

Spark devient plus pertinent quand le volume de données est important, notamment lorsqu'il est nécessaire de distribuer les traitements sur plusieurs machines et de les exécuter en parallèle. 

**Question 10** : Afficher la liste des colonnes de df_orders avec df.columns. Afficher les types avec df.dtypes. Identifier les
colonnes qui nécessitent un cast de type.

In [36]:
df_orders.columns

['order_id',
 'customer_id',
 'employee_id',
 'order_date',
 'required_date',
 'shipped_date',
 'ship_via',
 'freight',
 'ship_name',
 'ship_address',
 'ship_city',
 'ship_region',
 'ship_postal_code',
 'ship_country']

In [37]:
df_orders.dtypes

[('order_id', 'int'),
 ('customer_id', 'string'),
 ('employee_id', 'int'),
 ('order_date', 'date'),
 ('required_date', 'date'),
 ('shipped_date', 'date'),
 ('ship_via', 'int'),
 ('freight', 'double'),
 ('ship_name', 'string'),
 ('ship_address', 'string'),
 ('ship_city', 'string'),
 ('ship_region', 'string'),
 ('ship_postal_code', 'string'),
 ('ship_country', 'string')]

Les types des colonnes semblent avoir été correctement inférés par Spark. Les colonnes order_date, required_date et shipped_date ont été reconnues comme de type date. A première vue, aucun cast ne semble nécessaire.